python/cudf/cudf/core/column/categorical.py

```diff
@@ -281,9 +281,33 @@ def add_categories(self, new_categories, **kwargs):
         Categories (5, int64): [1, 2, 0, 3, 4]
         """
         new_categories = column.as_column(new_categories)
-        new_categories = self._column.categories.append(new_categories)
+        old_categories = self._column.categories
+
+        if (
+            old_categories.dtype == "object"
+            and new_categories.dtype != "object"
+        ) or (
+            new_categories.dtype == "object"
+            and old_categories.dtype != "object"
+        ):
+            raise TypeError(
+                "cudf does not support mixed types, please type-cast \
+                    new_categories to the same type as existing categories."
+            )
+        common_dtype = np.find_common_type(
+            [old_categories.dtype, new_categories.dtype], []
+        )
+
+        new_categories = new_categories.astype(common_dtype, copy=False)
+        old_categories = old_categories.astype(common_dtype, copy=False)
+
+        if old_categories.isin(new_categories).any():
+            raise ValueError("new categories must not include old categories")
+
+        new_categories = old_categories.append(new_categories)
         out_col = self._column
         if not self._categories_equal(new_categories, **kwargs):
+            kwargs["current_cats"] = old_categories
```

Comment:
kwargs are getting a bit overloaded here. Could we make a shallow copy of `self` and then change the categories on that and call `_set_categories` on that?

Reply:
Updated the `_set_categories` to actually take in current_categories as a parameter